opendir.c

```diff
@@ -47,143 +47,294 @@ struct watchman_dir_handle {
+  unused_parameter(canon_size);
+  errno = ENOSYS;
+  return -1;
 #endif
 }
 
-// Like lstat, but strict about symlinks in any path component
-int w_lstat(const char *path, struct stat *st) {
-#ifdef HAVE_OPENAT
-  char *pathcopy = strdup(path);
-  char *slash;
-  int dir_fd, res, err;
+/* Opens a file or directory, strictly prohibiting opening any symlinks
+ * in any component of the path, and strictly matching the canonical
+ * case of the file for case insensitive filesystems.
+ */
+static int open_strict(const char *path, int flags) {
+  int fd;
+  char canon[WATCHMAN_NAME_MAX];
+  int err = 0;
+  char *pathcopy = NULL;
+
+  if (strlen(path) >= sizeof(canon)) {
+    w_log(W_LOG_ERR, "open_strict(%s): path is larger than WATCHMAN_NAME_MAX\n",
+          path);
+    errno = EOVERFLOW;
+    return -1;
+  }
+
+  fd = open(path, flags);
+  if (fd == -1) {
+    return -1;
+  }
+
+  if (realpath_fd(fd, canon, sizeof(canon)) == 0) {
+    if (strcmp(canon, path) == 0) {
+      return fd;
+    }
+
+    w_log(W_LOG_DBG, "open_strict(%s): doesn't match canon path %s\n",
+        path, canon);
+    close(fd);
+    errno = ENOENT;
+    return -1;
+  }
+
+  if (errno != ENOSYS) {
+    err = errno;
+    w_log(W_LOG_ERR, "open_strict(%s): realpath_fd failed: %s\n", path,
+          strerror(err));
+    close(fd);
+    errno = err;
+    return -1;
+  }
 
+  // Fall back to realpath
+  pathcopy = w_realpath(path);
   if (!pathcopy) {
+    err = errno;
+    w_log(W_LOG_ERR, "open_strict(%s): realpath failed: %s\n", path,
+          strerror(err));
+    close(fd);
+    errno = err;
     return -1;
   }
 
-  slash = strrchr(pathcopy, '/');
-  if (!slash) {
+  if (strcmp(pathcopy, path)) {
+    // Doesn't match canonical case or the path we were expecting
+    w_log(W_LOG_ERR, "open_strict(%s): canonical path is %s\n",
+        path, pathcopy);
     free(pathcopy);
-    errno = EINVAL;
+    close(fd);
+    errno = ENOENT;
     return -1;
   }
+  free(pathcopy);
+  return fd;
+}
 
-  *slash = '\0';
-  dir_fd = open_dir_as_fd_no_symlinks(pathcopy, O_NOFOLLOW|O_CLOEXEC);
-  if (dir_fd == -1) {
-    free(pathcopy);
-    errno = ENOTDIR;
+// Like lstat, but strict about symlinks in any path component
+int w_lstat(const char *path, struct stat *st, bool case_sensitive) {
+  char *parent, *slash;
+  int fd = -1;
+  int err, res = -1;
+
+  parent = strdup(path);
+  if (!parent) {
     return -1;
   }
 
+  slash = strrchr(parent, '/');
+  if (slash) {
+    *slash = '\0';
+  }
+  fd = open_strict(parent, O_DIRECTORY | O_NOFOLLOW | O_CLOEXEC );
+  if (fd == -1) {
+    err = errno;
+    goto out;
+  }
+
   errno = 0;
-  res = fstatat(dir_fd, slash + 1, st, AT_SYMLINK_NOFOLLOW);
+#ifdef HAVE_OPENAT
+  res = fstatat(fd, slash + 1, st, AT_SYMLINK_NOFOLLOW);
+  err = errno;
+#else
+  res = lstat(path, st);
   err = errno;
+#endif
 
-  free(pathcopy);
-  close(dir_fd);
+  if (res == 0 && !case_sensitive && !is_basename_canonical_case(path)) {
+    err = ENOENT;
```

Comment:
Style: All the other blocks assign to `res` before `err`, why not do it here too?